### PR TITLE
Add at-or-insert to dict

### LINKED ISSUE #1182: Add a setdefault() method to dictionaries
### Description

Similar to Python dictionaries, Typst dictionaries could have a method of the form `setdefault(key, default)` that returns the value associated with `key` if it already exists, else inserts `(key, default)` and returns `default`. (The name is not really important)

### Use Case

When working with dictionaries, it is a common occurrence where you need to use a value that may already exist in the dictionary or needs to be inserted if it doesn't. Currently, doing this is somewhat verbose:

```typst
#let key = ...
#let value = if my-dict.keys().contains(key) {
  my-dict.at(key)
} else {
  let default = ...
  my-dict.insert(key, default)
  default
}
```

Obviously, it is trivial to write a utility function to achieve this functionality, but I think this use case is common enough to warrant a native function.

### PR DESCRIPTION
**Pull request by SekoiaTree**
Fixes #1182. Made off of the discussion in #2187. Adds "at-or-insert" to dict, which either gets the value in a dict if it exists, or inserts a given value (and returns it).

### CODE REVIEW

File: crates/typst/src/eval/dict.rs [hunk 1]
```diff
@@ -167,6 +167,20 @@ impl Dict {
         Arc::make_mut(&mut self.0).insert(key, value);
     }
 
+    /// Inserts a new pair into the dictionary and return the value. If the
```
Comment (laurmaedje) [inline crates/typst/src/eval/dict.rs:170]: The description is a bit strange. The first sentence sounds like it unconditionally inserts. It should start with the fact that it retrieves the value and then go on an explain what happens if the value wasn't present.

**Review by laurmaedje — COMMENTED** [review]
I'm still not sure whether this function feels Typst-y enough and whether giving up on `.at(key, insert: val)` because of the current implementation is the right way to go. I've asked for opinions on Discord.

**Comment by EpicEricEE** [comment]
Another idea would be to add an `update` flag to the `insert` method, and to make it return the current or new value.

**Comment by laurmaedje** [comment]
If we let insert return the value only if the `update` flag is passed, this would be a good alternative I think. @SekoiaTree What do you think?

File: crates/typst/src/eval/dict.rs [hunk 2]
```diff
@@ -167,6 +167,19 @@ impl Dict {
...
+    /// If the key does not exist, inserts the given value and returns it.
+    #[func]
+    pub fn at_or_insert(
+        &mut self,
+        /// The key at which to retrieve the item.
+        key: Str,
+        /// The value to insert if the key is not part of the dictionary.
+        value: Value,
+    ) -> Value {
+        Arc::make_mut(&mut self.0).entry(key).or_insert(value).clone()
```
Comment (laurmaedje) [inline crates/typst/src/eval/dict.rs:180]: Unconditionally making the `Arc` mutable might have more cost than `entry` is saving over double-lookup. Not sure.

**Comment by SekoiaTree** [comment]
So like, insert would return the old value (which I think is standard?) And if `replace` (clearer than update imo) is set to false, it won't replace the value and will return the old one? So
`insert(key, default-val, replace: false)` would be equivalent to `at-or-insert(key, default-val)`? Yeah, I don't mind that. Plus not replacing in some cases also fills the usecase of "insert if not present".

**Comment by laurmaedje** [comment]
> So like, insert would return the old value (which I think is standard?)

The problem with `insert` unconditionally returning the old value is that it might mess with joining. Especially if it defaults to `none` it might be non-obvious why joining fails sometimes (only when the key was already present). That's why I proposed only returning the value if replace is set to `false`, but that might also be pretty non-obvious. I'm not sure how to proceed.

**Comment by laurmaedje** [comment]
This seems pretty stuck at the moment. If we reach a consensus on a design, I'm happy to reopen it.

**Pull request closed without merging**